metrics_azurerm_reservation.go

```diff
@@ -0,0 +1,217 @@
+		prometheus.GaugeOpts{
+			Name: "azurerm_reservation_used_hours",
+			Help: "Azure ResourceManager Reservation Used Hours",
+		},
+		[]string{
+			"reservationOrderID",
+			"reservationID",
+			"skuName",
+			"kind",
+			"usageDate",
+		},
+	)
+
+	m.prometheus.reservationReservedHours = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "azurerm_reservation_reserved_hours",
+			Help: "Azure ResourceManager Reservation Reserved Hours",
+		},
+		[]string{
+			"reservationOrderID",
+			"reservationID",
+			"skuName",
+			"kind",
+			"usageDate",
+		},
+	)
+
+	m.prometheus.reservationTotalReservedQuantity = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "azurerm_reservation_total_reserved_quantity",
+			Help: "Azure ResourceManager Reservation Total Reserved Quantity",
+		},
+		[]string{
+			"reservationOrderID",
+			"reservationID",
+			"skuName",
+			"kind",
+			"usageDate",
+		},
+	)
+
+	m.Collector.RegisterMetricList("reservationInfo", m.prometheus.reservationInfo, true)
+	m.Collector.RegisterMetricList("reservationUsage", m.prometheus.reservationUsage, true)
+	m.Collector.RegisterMetricList("reservationMinUsage", m.prometheus.reservationMinUsage, true)
+	m.Collector.RegisterMetricList("reservationMaxUsage", m.prometheus.reservationMaxUsage, true)
+	m.Collector.RegisterMetricList("reservationUsedHours", m.prometheus.reservationUsedHours, true)
+	m.Collector.RegisterMetricList("reservationReservedHours", m.prometheus.reservationReservedHours, true)
+	m.Collector.RegisterMetricList("reservationTotalReservedQuantity", m.prometheus.reservationTotalReservedQuantity, true)
+}
+
+func (m *MetricsCollectorAzureRmReservation) Reset() {}
+
+func (m *MetricsCollectorAzureRmReservation) Collect(callback chan<- func()) {
+	m.collectReservationUsage(logger, callback)
+}
+
+func (m *MetricsCollectorAzureRmReservation) collectReservationUsage(logger *zap.SugaredLogger, callback chan<- func()) {
+	reservationInfo := m.Collector.GetMetricList("reservationInfo")
+	reservationUsage := m.Collector.GetMetricList("reservationUsage")
+	reservationMinUsage := m.Collector.GetMetricList("reservationMinUsage")
+	reservationMaxUsage := m.Collector.GetMetricList("reservationMaxUsage")
+	reservationUsedHours := m.Collector.GetMetricList("reservationUsedHours")
+	reservationReservedHours := m.Collector.GetMetricList("reservationReservedHours")
+	reservationTotalReservedQuantity := m.Collector.GetMetricList("reservationTotalReservedQuantity")
+
+	ctx := context.Background()
+	days := Config.Collectors.Reservation.FromDays
+	resourceScope := Config.Collectors.Reservation.ResourceScope
+	granularity := Config.Collectors.Reservation.Granularity
+
+	now := time.Now()
+	formattedDate := now.AddDate(0, 0, -days).Format("2006-01-02")
+	startDate := formattedDate
+	endDate := time.Now().Format("2006-01-02")
+
+	clientFactory, err := armconsumption.NewClientFactory("<subscription-id>", AzureClient.GetCred(), AzureClient.NewArmClientOptions())
```

Comment:
shouldn't `<subscription-id>` not set to the subscription id? and loop trough every subscription id in the collect function?

Reply:
Hi, thanx for your return,
It is surprising, but in fact the value should not be defined because this metrics_azurerm_reservation.go is intended to be used from an MCA tenant (which operates by billing profile rather than by subscription). Therefore, leaving "<subscription-id>" works well. We could rename the field but not set it to nil.